packages/optimizer/lib/transformers/ApplyCommonAttributes.js

```diff
@@ -0,0 +1,254 @@
+    }
+    return result;
+  }
+}
+
+/**
+ * Transforms a sizes attribute into CSS by creating media queries for each size.
+ */
+class SizesTransformer {
+  constructor() {
+    this.sizes = [];
+  }
+
+  transform(node, id) {
+    if (!node.attribs.srcset) {
+      // According to the Mozilla docs, a sizes attribute without a valid srcset attribute should have no effect.
+      // Therefore, it should simply be stripped, without producing media queries.
+      // @see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-sizes
+      return;
+    }
+    let sizes;
+    try {
+      sizes = parseSizes(node.attribs.sizes);
+    } catch (error) {
+      this.log.error(error.message);
+      // remove the sizes attribute as it's invalid anyway
+      return;
+    }
+    if (!sizes.defaultValue) {
+      // remove the sizes attribute as it's invalid anyway
+      return;
+    }
+    this.addSizes(id, sizes);
+  }
+
+  addSizes(id, sizes) {
+    this.sizes.push({
+      id,
+      defaultSize: sizes.defaultValue,
+      // The user agent will pick a width from the sizes attribute, using the first item with a <media-condition> (the part in parentheses) that evaluates to true.
+      // This means, we have to reverse the order the media queries in CSS to emulate this behavior (the last definition has precedence).
+      sizes: sizes.values.reverse(),
+    });
+  }
+
+  toString() {
+    let result = '';
+    for (const {sizes, defaultSize, id} of this.sizes) {
+      const selector = `#${id}`;
+      result += `${selector}{width:${defaultSize}}`;
+      for (const size of sizes) {
+        result += `@media ${size.media}{${selector}{width:${size.size}}}`;
+      }
+    }
+    return result;
+  }
+}
+
+/**
+ * Transforms a sizes attribute into CSS by creating media queries for each size.
+ */
+class HeightsTransformer {
+  constructor() {
+    this.heights = [];
+  }
+
+  transform(node, id) {
+    let heights;
+    try {
+      heights = parseSizes(node.attribs.heights);
+    } catch (error) {
+      this.log.error(error.message);
+      // remove the heights attribute as it's invalid anyway
+      return;
+    }
+    if (!heights.defaultValue) {
+      // remove the sizes attribute as it's invalid anyway
+      return;
+    }
+    this.addHeights(id, heights);
+  }
+
+  addHeights(id, heights) {
+    this.heights.push({
+      id,
+      defaultHeight: heights.defaultValue,
+      // The user agent will pick a width from the sizes attribute, using the first item with a <media-condition> (the part in parentheses) that evaluates to true.
+      // This means, we have to reverse the order the media queries in CSS to emulate this behavior (the last definition has precedence).
+      heights: heights.values.reverse(),
+    });
+  }
+
+  toString() {
+    let result = '';
+    for (const {heights, defaultHeight, id} of this.heights) {
+      const selector = `#${id}:first-child`;
```

Comment:
Shouldn't this be:

```suggestion
      const selector = `#${id}>:first-child`;
```

See https://github.com/ampproject/amp-wp/pull/4482#discussion_r421605106